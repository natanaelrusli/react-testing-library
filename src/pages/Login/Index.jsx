import { useState } from "react";
import {
  ErrorMessage,
  FormContainer,
  FormInput,
  FormItem,
  FormLabel,
  HeaderText,
  LoginButton,
  SignupLink
} from "./style";
import { Card } from '../../components/containers/Card'
import validator from "validator";

function Index() {
  const [loginInput, setLoginInput] = useState({
    email: "",
    password: "",
  });

  const [error, setError] = useState("");

  const handleChange = (e) => {
    setLoginInput({
      ...loginInput,
      [e.target.name]: e.target.value,
    });
  };

  const handleClick = (e) => {
    e.preventDefault();
    // JS validator
    // https://www.npmjs.com/package/validator
    if (!validator.isEmail(loginInput.email)) {
      return setError("The email you input is invalid");
    } else if (loginInput.password.length < 5) {
      return setError(
        "The password you entered should contain 5 or more characters"
      );
    } else if (loginInput.password !== loginInput.confirmPassword) {
      return setError("The passwords don't match. try again");
    }
  };

  return (
    <Card>
      <FormContainer>
        <HeaderText>Login</HeaderText>
        <FormItem>
          <FormLabel htmlFor="email">Email</FormLabel>
          <FormInput
            data-testid="emailInput"
            name="email"
            id="email"
            onChange={handleChange}
            value={loginInput.email}
          />
        </FormItem>
        <FormItem>
          <FormLabel htmlFor="password">Password</FormLabel>
          <FormInput
            data-testid="passwordInput"
            name="password"
            id="password"
            type={"password"}
            onChange={handleChange}
            value={loginInput.password}
          />
        </FormItem>
        <ErrorMessage>{error}</ErrorMessage>
        <LoginButton onClick={handleClick} className="pointer">
          Login
        </LoginButton>
        <p>Don't have an account? Sign Up <SignupLink href="/">here</SignupLink></p>
      </FormContainer>
    </Card>
  );
}

export default Index;
