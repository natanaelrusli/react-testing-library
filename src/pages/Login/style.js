import { Link } from 'react-router-dom'
import styled from 'styled-components'
import { theme } from '../../components/constants/Theme'

export const FormContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  margin-bottom: 20px;
  padding: 20px;
`

export const FormItem = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: start;
  margin-top: 5px;
  margin-bottom: 10px;
  width: 100%;
`

export const FormLabel = styled.p`
  font-size: 1.1rem;
  margin-bottom: 7px;
`

export const FormInput = styled.input`
  width: 100%;
  border-radius: 30px;
  padding: 5px;
  padding-left: 10px;
  padding-right: 10px;
  border: 1px solid black;
`

export const Container = styled.div`
  min-height: 100vh;
  display: flex;
  justify-content: center;
  align-items: center;
`

export const TextH1 = styled.h1`
  color: ${(props) => props.color || 'black'};
`

export const TextBig = styled(TextH1)`
  font-size: 4rem;
  text-transform: 'uppercase';
  &:hover {
    transform: scale(1.2);
    color: yellow;
    transition: 700ms;
  }
`

export const LinkCustom = styled(Link)`
  text-decoration: none;
`

export const Button = styled.button`
  color: black;
  background-color: cyan;
  font-size: 1em;
  margin: 1em;
  padding: 0.25em 1em;
  border: 2px solid blue;
  border-radius: 3px;
`

export const LoginButton = styled(Button)`
  background-color: #FFA500;
  color: white;
  border: none;
  outline: none;
  width: 100%;
  border-radius: 30px;
  padding: 10px;
  text-transform: uppercase;
  margin-top: 15px;
  transition: 500ms;
  &:hover {
    background-color: white;
    color: #FFA500;
    transition: 500ms;
    border: 1px solid #FFA500;
  }
`

export const GreenButton = styled(Button)`
  color: green;
  border-color: yellow;
`

export const ErrorMessage = styled.p`
  text-align: left;
  color: red;
`

export const SignupLink = styled.a`
  color: black;
  text-decoration: none;
  cursor: pointer;
  color: #FFA500;
`

export const HeaderText = styled.h1`
  color: ${theme.colors.orange};
  margin-bottom: 20px;
  margin-top: 20px;
  text-align: center;
`