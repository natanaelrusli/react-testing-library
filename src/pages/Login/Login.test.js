import Login from './Index'
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer'

test('inputs should be initially empty', () => {
  render(<Login />)

  expect(screen.getByTestId('emailInput').value).toBe('')
  expect(screen.getByTestId('passwordInput').value).toBe('')
})

describe('snapshot testing', () => {
  test('snapshot for login component', () => {
    const login = renderer.create(<Login />).toJSON();
    expect(login).toMatchSnapshot()
  })
})