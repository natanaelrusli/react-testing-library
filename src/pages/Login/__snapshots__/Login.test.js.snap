// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing snapshot for login component 1`] = `
<div
  className="sc-bcXHqe cFtqiI"
>
  <div
    className="sc-gswNZR fiWvVF"
  >
    <p
      className="sc-dkrFOg gboSJr"
      htmlFor="email"
    >
      Email
    </p>
    <input
      className="sc-hLBbgP dnwCP"
      data-testid="emailInput"
      id="email"
      name="email"
      onChange={[Function]}
      value=""
    />
  </div>
  <div
    className="sc-gswNZR fiWvVF"
  >
    <p
      className="sc-dkrFOg gboSJr"
      htmlFor="password"
    >
      Password
    </p>
    <input
      className="sc-hLBbgP dnwCP"
      data-testid="passwordInput"
      id="password"
      name="password"
      onChange={[Function]}
      type="password"
      value=""
    />
  </div>
  <p
    className="sc-kDvujY hbSaTb"
  />
  <button
    className="sc-ftTHYK sc-pyfCe koensh fTUoFM pointer"
    onClick={[Function]}
  >
    Login
  </button>
  <p>
    Don't have an account? Sign Up 
    <a
      className="sc-ipEyDJ dNNBrC"
      href="/"
    >
      here
    </a>
  </p>
</div>
`;
